CMakeLists.txt

```diff
@@ -2,156 +2,243 @@
+    CACHE INTERNAL "")
+set(PYTHON_MODULE_PREFIX
+    ${PYTHON_MODULE_PREFIX}
+    CACHE INTERNAL "")
+set(PYTHON_MODULE_EXTENSION
+    ${PYTHON_MODULE_EXTENSION}
+    CACHE INTERNAL "")
+set(PYTHON_VERSION_MAJOR
+    ${PYTHON_VERSION_MAJOR}
+    CACHE INTERNAL "")
+set(PYTHON_VERSION_MINOR
+    ${PYTHON_VERSION_MINOR}
+    CACHE INTERNAL "")
+set(PYTHON_IS_DEBUG
+    "${PYTHON_IS_DEBUG}"
+    CACHE INTERNAL "")
+
+if(PYBIND11_TEST OR (BUILD_TESTING AND PYBIND11_MASTER_PROJECT))
   add_subdirectory(tests)
 endif()
 
-include(GNUInstallDirs)
-include(CMakePackageConfigHelpers)
-
-# extract project version from source
-file(STRINGS "${PYBIND11_INCLUDE_DIR}/pybind11/detail/common.h" pybind11_version_defines
-     REGEX "#define PYBIND11_VERSION_(MAJOR|MINOR|PATCH) ")
-foreach(ver ${pybind11_version_defines})
-  if (ver MATCHES "#define PYBIND11_VERSION_(MAJOR|MINOR|PATCH) +([^ ]+)$")
-    set(PYBIND11_VERSION_${CMAKE_MATCH_1} "${CMAKE_MATCH_2}" CACHE INTERNAL "")
-  endif()
-endforeach()
-set(${PROJECT_NAME}_VERSION ${PYBIND11_VERSION_MAJOR}.${PYBIND11_VERSION_MINOR}.${PYBIND11_VERSION_PATCH})
-message(STATUS "pybind11 v${${PROJECT_NAME}_VERSION}")
+if(USE_PYTHON_INCLUDE_DIR)
+  file(RELATIVE_PATH CMAKE_INSTALL_INCLUDEDIR ${CMAKE_INSTALL_PREFIX} ${PYTHON_INCLUDE_DIRS})
+endif()
 
-option (USE_PYTHON_INCLUDE_DIR "Install pybind11 headers in Python include directory instead of default installation prefix" OFF)
-if (USE_PYTHON_INCLUDE_DIR)
-    file(RELATIVE_PATH CMAKE_INSTALL_INCLUDEDIR ${CMAKE_INSTALL_PREFIX} ${PYTHON_INCLUDE_DIRS})
+# Note: when creating targets, you cannot use if statements at configure time - you need generator
+# expressions, because those will be placed in the target file.
+
+# Build an interface library target:
+add_library(pybind11 INTERFACE)
+add_library(pybind11::pybind11 ALIAS pybind11) # to match exported target
+
+target_include_directories(
+  pybind11 ${pybind11_system} INTERFACE $<BUILD_INTERFACE:${PYBIND11_INCLUDE_DIR}>
+                                        $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>)
+# Only add Python for build - must be added during the import for config since it has to be
+# re-discovered.
+target_include_directories(pybind11 SYSTEM INTERFACE $<BUILD_INTERFACE:${PYTHON_INCLUDE_DIRS}>)
+
+if(CMAKE_VERSION VERSION_LESS 3.13)
+  target_compile_features(
+    pybind11
+    INTERFACE cxx_auto_type
+              cxx_constexpr
+              cxx_defaulted_functions
+              cxx_lambdas
+              cxx_nullptr
+              cxx_range_for
+              cxx_right_angle_brackets
+              cxx_strong_enums
+              cxx_user_literals
+              cxx_variadic_templates)
+else()
+  # This was added in CMake 3.8, but we are keeping a consistent breaking point for the config file
+  # at 3.13. A config generated by CMake 3.13+ can ony be read in 3.13+ due to the SHELL usage
+  # later, so this is safe to do.
+  target_compile_features(pybind11 INTERFACE cxx_std_11)
```

Comment:
Regarding this line: how will this work when the user wants to use newer versions of the standard in their library? Does CMake somehow take the max(..) of the `INTERFACE` value?

Reply:
`target_compile_features` says "this needs at least the features of C++11" to compile". If your compiler doesn't support the feature, you get a nice, configure time error (more applicable to the non-meta feature method). If you have a target with multiple features, it selects the flag needed to get all of them (so yes, basically the max). If you have a `CXX_STANDARD` set, then that is used for the flag, or (again) a nice configure time error is given you try to set a standard too low.

In CMake 3.11+, you can set COMPILE_OPTIONS on individual files!